### PR TITLE
geo/geomfn: add point in polygon optimization to st_intersects

### PR DESCRIPTION
**Pull request by andyyang890**
This patch improves the performance of st_intersects for the
common use case of testing whether a (multi)polygon and a
(multi)point intersect.

Release note: None

### CODE REVIEW

**Comment by andyyang890** [comment]
In my local testing, this improved the performance of the benchmark query from https://blog.cleverelephant.ca/2018/09/postgis-external-storage.html from ~55s to ~17s.

**Comment by otan** [comment]
> In my local testing, this improved the performance of the benchmark query from https://blog.cleverelephant.ca/2018/09/postgis-external-storage.html from ~55s to ~17s.

nice! do you have a profile of where time is being spent now?

**Comment by andyyang890** [comment]
> > In my local testing, this improved the performance of the benchmark query from https://blog.cleverelephant.ca/2018/09/postgis-external-storage.html from ~55s to ~17s.
> 
> nice! do you have a profile of where time is being spent now?

https://share.polarsignals.com/f694a64/

**Review by otan — APPROVED** [review]
i would see if you can make the same optimisations for st_contains and st_within

**Comment by andyyang890** [comment]
> i would see if you can make the same optimisations for st_contains and st_within

I used the same baseline query but changed the function to ST_Contains and ST_Within and the performance of both improves from ~55s to ~12s.

Profile for ST_Contains: https://share.polarsignals.com/7a3ba6f/

File: pkg/geo/geomfn/distance.go [hunk 1]
```diff
@@ -814,3 +830,149 @@ func verifyDensifyFrac(f float64) error {
...
+				return false, err
+			}
+			if !hasPolygon {
+				break
+			}
+			polygonGeometry, err := geo.MakeGeometryFromGeomT(polygon)
+			if err != nil {
+				return false, err
+			}
+			intersects, err := PointIntersectsPolygon(pointGeometry, polygonGeometry)
```
Comment (otan) [inline pkg/geo/geomfn/distance.go:894]: I don't think within / contains is a strict intersection -- it *cannot* be on any of the edges.

see https://mentin.medium.com/which-predicate-cb608b470471.

**Review by otan — CHANGES_REQUESTED** [review]
Nice.

I've added a few comments to things I've noticed which could improve performance which I missed in the earlier PR.

You should probably have merged that intersects one first and then did this on top, to break up the work a bit! It's okay to break this stuff up and take care of perfection separately. This PR got more difficult factoring in Contains and WIthin and as a result requires cognitive load on top of the first PR to review. Ah well.

File: pkg/geo/geomfn/distance.go (hunk 1 above)
Comment (otan) [reply pkg/geo/geomfn/distance.go:894]: (i would say if you can, do within/contains in a separate PR. you can keep the tests -- they should still pass)

File: pkg/geo/geomfn/distance.go [hunk 2]
```diff
@@ -814,3 +830,149 @@ func verifyDensifyFrac(f float64) error {
...
+			return false, nil
+		}
+	}
+	return ret, nil
+}
+
+// PointIntersectsPolygon returns whether a point intersects with a polygon.
+func PointIntersectsPolygon(point geo.Geometry, polygon geo.Geometry) (bool, error) {
+	// Convert point from a geo.Geometry to a *geodist.Point.
+	pointT, err := point.AsGeomT()
```
[the remark this answers is not in the document]
Comment (andyyang890) [reply pkg/geo/geomfn/distance.go:921]: Good point, will change. I think the reason I made this choice earlier was that I had wanted this function to be called directly by `func Intersects(a geo.Geometry, b geo.Geometry) (bool, error)`.

File: pkg/geo/geomfn/binary_predicates.go [hunk 3]
```diff
@@ -225,6 +246,26 @@ func Overlaps(a geo.Geometry, b geo.Geometry) (bool, error) {
 	return geos.Overlaps(a.EWKB(), b.EWKB())
 }
 
+// PointKindAndPolygonKind returns whether a pair of geometries contains
+// a (multi)point and a (multi)polygon. It is used to determine if the
+// point in polygon optimization can be applied.
+func PointKindAndPolygonKind(
+	a geo.Geometry, b geo.Geometry,
+) (validPair bool, pointKind *geo.Geometry, polygonKind *geo.Geometry) {
+	for _, geomPtr := range []*geo.Geometry{&a, &b} {
```
[the remark this answers is not in the document]
Comment (otan) [reply pkg/geo/geomfn/binary_predicates.go:255]: the slice is not reference, so the struct is only as big as the size of the SpatialObject struct, which isn't too bad.

we had pointers originally and it caused a lot of allocations -- https://github.com/cockroachdb/cockroach/pull/53446

File: pkg/geo/geomfn/distance.go (hunk 1 above)
Comment (otan) [reply pkg/geo/geomfn/distance.go:894]: yes, but as commented earlier, please add this in a separate PR!

Comment (otan) [reply pkg/geo/geomfn/distance.go:894]: http://lin-ear-th-inking.blogspot.com/2007/06/subtleties-of-ogc-covers-spatial.html

i need to read this every time, lol

File: pkg/geo/geomfn/binary_predicates.go [hunk 4]
```diff
@@ -225,6 +233,27 @@ func Overlaps(a geo.Geometry, b geo.Geometry) (bool, error) {
...
+		switch geometry.ShapeType2D() {
+		case geopb.ShapeType_Point, geopb.ShapeType_MultiPoint:
+			validPoint = true
+			pointKind = geometry
+		case geopb.ShapeType_Polygon, geopb.ShapeType_MultiPolygon:
+			validPolygon = true
+			polygonKind = geometry
+		}
+	}
+	validPair = validPoint && validPolygon
```
Comment (otan) [inline pkg/geo/geomfn/binary_predicates.go:253]: nit: i prefer explicit returns, i.e. `return validPoint && validPolygon, pointKind, polygonKind`

File: pkg/geo/geomfn/binary_predicates_test.go [hunk 5]
```diff
@@ -28,6 +29,18 @@ var (
...
 	middleLine           = geo.MustParseGeometry("LINESTRING(-0.5 0.5, 0.5 0.5)")
+	leftRectWithHole     = geo.MustParseGeometry("POLYGON((-1.0 0.0, 0.0 0.0, 0.0 1.0, -1.0 1.0, -1.0 0.0), " +
+		"(-0.75 0.25, -0.75 0.75, -0.25 0.75, -0.25 0.25, -0.75 0.25))")
+	bothLeftRects = geo.MustParseGeometry("MULTIPOLYGON(((-1.0 0.0, 0.0 0.0, 0.0 1.0, -1.0 1.0, -1.0 0.0)), " +
+		"((-1.0 0.0, 0.0 0.0, 0.0 1.0, -1.0 1.0, -1.0 0.0), (-0.75 0.25, -0.75 0.75, -0.25 0.75, -0.25 0.25, -0.75 0.25)))")
+	bothLeftRectsHoleFirst = geo.MustParseGeometry("MULTIPOLYGON(" +
+		"((-1.0 0.0, 0.0 0.0, 0.0 1.0, -1.0 1.0, -1.0 0.0), " +
+		"(-0.75 0.25, -0.75 0.75, -0.25 0.75, -0.25 0.25, -0.75 0.25)), " +
+		"((-1.0 0.0, 0.0 0.0, 0.0 1.0, -1.0 1.0, -1.0 0.0)))")
+	leftRectCornerPoint     = geo.MustParseGeometry("POINT(-1.0 0.0)")
```
Comment (otan) [inline pkg/geo/geomfn/binary_predicates_test.go:40]: do you have `leftRectEdgePoint`?

**Review by otan — APPROVED** [review]

**Comment by andyyang890** [comment]
Having `PointKindAndPolygonKind` return values instead of pointers seems to have improved the performance further to ~10s.

Profile: https://share.polarsignals.com/ec7ae37/

**Comment by andyyang890** [comment]
bors r=otan

File: pkg/geo/geomfn/distance.go (hunk 1 above)
Comment (andyyang890) [reply pkg/geo/geomfn/distance.go]: Added the PIP optimization to st_covers in #62582

**Pull request merged**